mailpoet/lib/API/MP/v1/Segments.php

```diff
@@ -27,31 +39,125 @@ public function getAll(): array {
+    } catch (\Exception $e) {
       throw new APIException(
-        __('List name is required.', 'mailpoet'),
-        APIException::LIST_NAME_REQUIRED
+        __('The list couldn’t be created in the database', 'mailpoet'),
+        APIException::FAILED_TO_SAVE_LIST
       );
     }
 
-    if (!$this->segmentsRepository->isNameUnique($data['name'], null)) {
+    return $this->buildItem($segment);
+  }
+
+  public function updateList(array $data): array {
+    // firstly validation on list id
+    $this->validateSegmentId((string)($data['id'] ?? ''));
+
+    // secondly validation on list name
+    $this->validateSegmentName($data);
+
+    try {
+      $segment = $this->segmentsRepository->createOrUpdate(
+        $data['name'],
+        $data['description'] ?? '',
+        SegmentEntity::TYPE_DEFAULT,
+        [],
+        $data['id']
+      );
+    } catch (\Exception $e) {
       throw new APIException(
-        __('This list already exists.', 'mailpoet'),
-        APIException::LIST_EXISTS
+        __('The list couldn’t be updated in the database', 'mailpoet'),
+        APIException::FAILED_TO_UPDATE_LIST
+      );
+    }
+
+    return $this->buildItem($segment);
+  }
+
+  public function deleteList(string $listId): bool {
+    $this->validateSegmentId($listId);
+
+    $activelyUsedNewslettersSubjects = $this->newsletterSegmentRepository->getSubjectsOfActivelyUsedEmailsForSegments([$listId]);
+    if (isset($activelyUsedNewslettersSubjects[$listId])) {
+      throw new APIException(
+        str_replace(
+          '%1$s',
+          "'" . join("', '", $activelyUsedNewslettersSubjects[$listId] ) . "'",
+          // translators: %1$s is a comma-seperated list of emails for which the segment is used.
+          _x('List cannot be deleted because it’s used for %1$s email', 'Alert shown when trying to delete segment, which is assigned to any automatic emails.', 'mailpoet')
+        ),
+        APIException::LIST_USED_IN_EMAIL
+      );
+    }
+
+    $activelyUsedFormNames = $this->formsRepository->getNamesOfFormsForSegments();
+    if (isset($activelyUsedFormNames[$listId])) {
+      throw new APIException(
+        str_replace(
+          '%1$s',
+          "'" . join("', '", $activelyUsedFormNames[$listId] ) . "'",
+          // translators: %1$s is a comma-seperated list of forms for which the segment is used.
+          _nx(
+            'List cannot be deleted because it’s used for %1$s form',
+            'List cannot be deleted because it’s used for %1$s forms',
+            count($activelyUsedFormNames[$listId]),
+            'Alert shown when trying to delete segment, when it is assigned to a form.',
+            'mailpoet'
+          )
+        ),
+        APIException::LIST_USED_IN_FORM
       );
     }
 
     try {
-      $segment = $this->segmentsRepository->createOrUpdate($data['name'], $data['description'] ?? '');
+      $this->segmentsRepository->bulkDelete([$listId]);
```

Comment:
If I try to delete a list that can not be deleted, because type !== 'default', ex: 'WooCommerce Customers', the list is not deleted but the API returns true.

One option would be to the check the return of `$this->segmentsRepository->bulkDelete([$listId]);` 
if > 0 return true, 
if not, throw APIException::FAILED_TO_DELETE_LIST (or a generic Exception that would convert into APIException::FAILED_TO_DELETE_LIST on the catch block)

What do you think? 

You can't delete dynamic segments, is this expected?

Reply:
Thank you, Brezo! Nice catch. MP API should allow default lists to be deleted because it only supports creating default lists.